srml/slashing/src/lib.rs

```diff
@@ -0,0 +1,108 @@
+
+//! Abstract slashing interface
+//!
+//! That gives functionality to specialize slashing and misconduct for a given type
+//! In order to customize severity level and misconduct fees.
+//!
+//! TODO(niklasad1): provide default impl?
+//! TODO(niklasad1): move the `srml/support/src/traits.rs`?
+//! TODO(niklasad1): chain reversion are ignored, do we need specific method/function for that?
+//!
+//! For example an use case could be to increase severity level exponentially on concurrent culprits
+
+/// # Examples
+///	```
+///	use srml_slashing::{Slashing, Misconduct};
+/// use parity_codec::Codec;
+///	use sr_primitives::traits::{SimpleArithmetic, MaybeSerializeDebug};
+///
+///	// Dummy example that represent severity level as a simple counter that gets reset `on_signal`
+///	// Slashes `balance / severity`
+///	pub struct MyType;
+///
+/// impl Slashing for MyType {
+///		type AccountId = u64;
+///		type Balance = u32;
+///		type Severity = u32;
+///
+///		fn on_slash(
+///			severity: Self::Severity,
+///			who: Self::AccountId,
+///			balance: Self::Balance,
+///			misconduct: impl Misconduct<Balance = Self::Balance>
+///		) -> (Self::Balance, Self::Severity) {
+///			let new_balance = misconduct.on_misconduct(balance, severity);
+///			(new_balance, severity + 1)
+///		}
+///
+///		fn on_signal(severity: Self::Severity) -> Self::Severity {
+///			0
+///		}
+///	}
+///
+///	enum MyTypeMisconduct {
+///		A,
+///		B,
+///		C,
+///	}
+///
+///	impl Misconduct for MyTypeMisconduct {
+///		type Balance = u32;
+///
+///		fn on_misconduct<B, S>(&self, balance: B, severity: S) -> Self::Balance
+///		where
+///			B: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default + Into<Self::Balance>,
+///			S: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default + Into<Self::Balance>
+///		{
+///			let x: Self::Balance = balance.into() / severity.into();
+///
+///			match &self {
+///				MyTypeMisconduct::A => x + 0,
+///				MyTypeMisconduct::B => x + 1,
+///				MyTypeMisconduct::C => x + 2,
+///			}
+///		}
+///	}
+/// ```
+
+use parity_codec::Codec;
+use sr_primitives::traits::{SimpleArithmetic, MaybeSerializeDebug};
+
+/// Represents `generic` misconduct to be slashed
+pub trait Misconduct {
+	/// Account balance
+	type Balance: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default;
+
+	/// Calculate new balance based on the `misconduct` and `severity`
+	fn on_misconduct<B, S>(&self, balance: B, severity: S) -> Self::Balance
+	where
+		B: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default + Into<Self::Balance>,
+		S: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default + Into<Self::Balance>;
+}
+
+/// Slashing interface
+// TODO(niklasad1): should imbalance be used here such as `trait::Currency`?!
+pub trait Slashing {
+	/// Account id
+	type AccountId: Codec + Copy + MaybeSerializeDebug + Default;
+	/// Account balance
+	type Balance: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default;
```

Comment:
One thing we could do is provide an intermediate layer between the slashing module and the thing actually being slashed. So the slashing trait shouldn't try to deal in balances or anything like that but only in severity. Then there will be another trait which `Slashing` can invoke (`trait DoSlash`?), which would act as a bridge between the misbehavior aggregation logic (this module) and actually slashing (balances).

IMO that would make things a bit more testable because we can write small tests of `srml_slashing` to see if the right severity is being reported. Then when we have a `DoSlash` implementation which is also very testable.

Reply:
with that in mind, maybe this would be the `slash_aggregator` module?

Reply:
Yeah, I agree it could be easier to use/understand and to test if we could decouple `severity` and `balance`.